Database/MySQL/Simple/QueryResults/Generic.hs

```diff
@@ -0,0 +1,88 @@
+{-| Generic deriviation of 'Database.MySQL.Simple.QueryResults.QueryResults'. -}
+{-# LANGUAGE
+    ScopedTypeVariables
+  , TypeOperators
+  , InstanceSigs
+  , FlexibleContexts
+  , AllowAmbiguousTypes
+  #-}
+{-# OPTIONS_GHC -Wall -Werror #-}
+
+module Database.MySQL.Simple.QueryResults.Generic
+  ( QueryResults(..)
+  , convert
+  ) where
+
+import Prelude ()
+import Database.MySQL.Simple.Prelude
+import Data.ByteString (ByteString)
+import GHC.Generics
+import qualified GHC.Generics as Generics
+import Control.Exception (throw)
+
+import Database.MySQL.Base.Types (Field)
+import Database.MySQL.Simple.Result
+  (Result, ResultError)
+import qualified Database.MySQL.Simple.Result as Result
+import Data.Proxy (Proxy(Proxy))
+import Database.MySQL.Simple.Arity (Arity, arity, KnownNat)
+
+-- | Generic implementation of
+-- 'Database.MySQL.Simple.QueryResults.Generic.convertResults'.
+convert
+  :: forall a
+  .  Generic a
+  => KnownNat (Arity (Rep a))
+  => QueryResults (Rep a)
+  => [Field]
+  -> [Maybe ByteString]
+  -> a
+convert xs ys = Generics.to $ convertResults err xs ys
+  where
+  err = Result.convertException xs ys $ fromIntegral $ arity (Proxy :: Proxy a)
+
+-- | The generic counterpart to 'Database.MySQL.Simple.QueryResults.QueryResults'.
+class QueryResults f where
+  convertResults :: ResultError -> [Field] -> [Maybe ByteString] -> f a
+
+-- This instance might not make sense, though the signature of
+-- 'convertResults' sort of implies that it does since it takes in two
+-- '[]''s in stead of two 'NonEmpty''s.
+instance QueryResults U1 where
+  convertResults err xs ys = case zip xs ys of
+    [] -> U1
+    _  -> throw err
+
+instance Result a => QueryResults (K1 i a) where
+  convertResults err xs ys = case zip xs ys of
+    [(x, y)] -> K1 $ Result.convert x y
+    _ -> throw err
+
+instance QueryResults a => QueryResults (M1 i c a) where
+  convertResults err xs ys =
+    M1 $ convertResults err xs ys
+
+instance (QueryResults a, QueryResults b) => QueryResults (a :*: b) where
+  convertResults
+    :: forall x
+    .  ResultError
+    -> [Field]
+    -> [Maybe ByteString]
+    -> (:*:) a b x
+  convertResults err (x:xs) (y:ys) =
+    -- I'm concerned about this implementation since it's biased
+    -- towards one side.  Meaning that it's expecting the chain of
+    -- ':*:' to lean to the right.  As an example of the problem
```

Comment:
I'm not clear on how this all fits together, but that's an extremely bad assumption. Once there are more than three fields, the generic representation will be a tree.

Reply:
Yeah I should look into this.

Reply:
I didn't manage to describe the problem properly in this comment. The issue is not what the generic representation generated by `from` is. The problem is the generic representation that *we* generate and then later try and convert back to some type using `to`. Although the problem may be related since I suspect `from` and `to` are inverses.

Reply:
I can't really figure out how to give a sensible definition for this. If someone wants to take a stab at it this snippet may come in handy for testing. https://gist.github.com/fredefox/79e8bcae045eada98934df88bdad7783

Reply:
I've added a test-case for this.